File: packages/domain/src/matching.ts
/**
 * Which load, and whose bid.
 *
 * Tracking is the wedge; this is the business. Two rankings, pointing in
 * opposite directions:
 *
 * - `rankLoads` — what a carrier should take next;
 * - `rankBids` — whose offer a shipper should accept.
 *
 * Both return every candidate, scored and with the score explained, rather
 * than a filtered shortlist. A ranking that silently drops options is one the
 * user cannot argue with, and the first thing a haulier does with a
 * recommendation is argue with it.
 */

import { distance, type Metres, type Position } from './geo.ts';
import { quote, type TruckClass, type Tonnes, fits } from './pricing.ts';
import { type Kobo } from './money.ts';

export interface Load {
  readonly id: string;
  readonly origin: Position;
  readonly destination: Position;
  readonly weight: Tonnes;
  readonly requires: TruckClass;
  /** What the shipper is offering, or undefined if it is open to bids. */
  readonly offered?: Kobo;
  /** Collection must start by this time. */
  readonly readyBy: Date;
  readonly expiresAt: Date;
}

export interface Carrier {
  /** Where the truck is, or will be when it is free. */
  readonly at: Position;
  /** When it is free. */
  readonly freeFrom: Date;
  readonly truck: TruckClass;
  /**
   * Where the truck is trying to get back to.
   *
   * This is the whole product. An empty truck running 900 km home earns
   * nothing and burns diesel the entire way; a load going that direction at
   * half price is better than a full-price load going the wrong way. Without a
   * base, this reduces to ordinary proximity matching.
   */
  readonly base?: Position;
}

/**
 * Why a load did not score well — or could not be taken at all.
 *
 * `blocked` is the important distinction: a load that cannot physically be
 * taken is shown greyed with the reason, not hidden. A carrier who cannot see
 * why the 30-tonne load is not on their list assumes the app is broken.
 */
export type Blocker = 'too_heavy' | 'wrong_class' | 'expired' | 'cannot_reach';

export interface LoadScore {
  readonly load: Load;
  /** 0–1. Meaningless in isolation; the ordering is the product. */
  readonly score: number;
  readonly blocked: Blocker | null;
  /** Empty kilometres to reach the pickup. */
  readonly deadhead: Metres;
  /** Metres of the run home this load covers. Negative means it takes the
   * truck further away. */
  readonly progressHome: Metres;
  /** One line, rendered under the load on the carrier's list. */
  readonly because: string;
}

/** Deadhead beyond this is not worth scoring against; the load is not local. */
export const MAX_DEADHEAD_M = 400_000;

/**
 * Ranks available loads for a carrier.
 *
 * Sorted best first. Blocked loads sort last regardless of score, because a
 * load that cannot be taken should never sit above one that can, however
 * attractive it looks.
 */
export function rankLoads(
  carrier: Carrier,
  loads: readonly Load[],
  now: Date,
): readonly LoadScore[] {
  return loads
    .map((load) => scoreLoad(carrier, load, now))
    .sort((a, b) => {
      if ((a.blocked === null) !== (b.blocked === null)) {
        return a.blocked === null ? -1 : 1;
      }
      return b.score - a.score;
    });
}

function scoreLoad(carrier: Carrier, load: Load, now: Date): LoadScore {
  const deadhead = distance(carrier.at, load.origin);
  const haul = distance(load.origin, load.destination);

  const progressHome =
    carrier.base === undefined
      ? 0
      : distance(carrier.at, carrier.base) -
        distance(load.destination, carrier.base);

  const blocked = blockerFor(carrier, load, deadhead, now);
  if (blocked !== null) {
    return {
      load,
      score: 0,
      blocked,
      deadhead,
      progressHome,
      because: explainBlocker(blocked),
    };
  }

  // Three things decide it, and the weights say which matters:
  //
  //   value    what the trip pays against what it costs to reach
  //   homeward how much of the empty run home this load covers
  //   urgency  how soon it has to be collected
  //
  // Homeward is weighted almost as heavily as value on purpose. It is the
  // asymmetry the product is named after, and a matcher that treats a return
  // load as just another load is a load board.
  const value = valueScore(load, haul, deadhead);
  const homeward = homewardScore(progressHome, haul, carrier.base !== undefined);
  const urgency = urgencyScore(load, now);

  const score = clamp(0.45 * value + 0.4 * homeward + 0.15 * urgency);

  return { load, score, blocked: null, deadhead, progressHome, because: explain(deadhead, progressHome, carrier.base !== undefined) };
}

function blockerFor(
  carrier: Carrier,
  load: Load,
  deadhead: Metres,
  now: Date,
): Blocker | null {
  if (load.expiresAt.getTime() <= now.getTime()) return 'expired';
  if (!fits(carrier.truck, load.weight)) return 'too_heavy';
  if (carrier.truck !== load.requires) return 'wrong_class';
  if (deadhead > MAX_DEADHEAD_M) return 'cannot_reach';
  return null;
}

/**
 * Paid distance against total distance.
 *
 * A 500 km haul reached by 50 km of empty running is 91% productive. The same
 * haul reached by 400 km of empty running is 56%, and no rate makes that up.
 */
function valueScore(load: Load, haul: Metres, deadhead: Metres): number {
  const total = haul + deadhead;
  if (total === 0) return 0;
  const productive = haul / total;

  // A shipper offering above the indicative range is worth noticing; one
  // offering below it is worth noticing too, in the other direction.
  const indicative = quote(load.requires, haul).mid;
  const premium =
    load.offered === undefined || indicative === 0
      ? 1
      : clamp(load.offered / indicative, 0.5, 1.5);

  return clamp(productive * premium);
}

/**
 * How much of the run home this load covers, as a fraction of its own length.
 *
 * A load that goes exactly the right way scores 1. One that goes sideways
 * scores 0.5 — neutral, not punished, because a paying sideways load still
 * beats an empty truck. One that goes backwards scores toward 0.
 *
 * With no base, this returns a flat 0.5 rather than 0: a carrier who has not
 * told us where home is should not have every load marked down.
 */
function homewardScore(
  progressHome: Metres,
  haul: Metres,
  hasBase: boolean,
): number {
  if (!hasBase || haul === 0) return 0.5;
  return clamp(0.5 + progressHome / (2 * haul));
}

/** Loads that must move today outrank loads that can wait a week. */
function urgencyScore(load: Load, now: Date): number {
  const hours = (load.readyBy.getTime() - now.getTime()) / 3_600_000;
  if (hours <= 0) return 1;
  if (hours >= 72) return 0;
  return 1 - hours / 72;
}

function explain(deadhead: Metres, progressHome: Metres, hasBase: boolean): string {
  const empty = `${Math.round(deadhead / 1000)} km empty to the pickup`;
  if (!hasBase) return `${empty}.`;
  if (progressHome > 50_000) {
    return `${empty}, and it covers ${Math.round(progressHome / 1000)} km of the run home.`;
  }
  if (progressHome < -50_000) {
    return `${empty}, but it takes you ${Math.round(-progressHome / 1000)} km further from base.`;
  }
  return `${empty}; neither toward base nor away from it.`;
}

function explainBlocker(blocked: Blocker): string {
  switch (blocked) {
    case 'too_heavy':
      return 'Heavier than your truck carries.';
    case 'wrong_class':
      return 'The shipper asked for a different class of truck.';
    case 'expired':
      return 'This load has expired.';
    case 'cannot_reach':
      return `More than ${MAX_DEADHEAD_M / 1000} km of empty running away.`;
  }
}

export interface Bid {
  readonly id: string;
  readonly carrierId: string;
  readonly amount: Kobo;
  /** Completed trips, and how many of them arrived when promised. */
  readonly tripsCompleted: number;
  readonly tripsOnTime: number;
  /** Where the truck is now. */
  readonly at: Position;
  readonly placedAt: Date;
}

export interface BidScore {
  readonly bid: Bid;
  readonly score: number;
  /** 0–1, or null when the carrier has too little history to have one. */
  readonly reliability: number | null;
  readonly kmToPickup: number;
  readonly because: string;
}

/** Below this many completed trips, a carrier has no meaningful record. */
export const MINIMUM_TRIPS_FOR_RELIABILITY = 5;

/**
 * How far above the cheapest bid a carrier may ask before price scores zero.
 *
 * A quarter. A carrier asking 25% more than the cheapest offer had better be
 * winning on record alone, and at that point the shipper should be looking at
 * the two figures themselves rather than at a ranking.
 */
export const PREMIUM_TOLERANCE = 0.25;

/**
 * Ranks bids for a shipper.
 *
 * The cheapest bid is not the best bid, and this is where the product either
 * earns trust or loses it. A carrier with forty on-time trips asking 10% more
 * than one with none is the better answer, and the ranking says so — but it
 * shows the price and the record side by side so the shipper can overrule it.
 *
 * A new carrier is not scored as unreliable. They are scored as *unknown*,
 * which sits between good and bad rather than at the bottom: a marketplace
 * that never surfaces a new carrier never gets a second one.
 */
export function rankBids(bids: readonly Bid[], pickup: Position): readonly BidScore[] {
  if (bids.length === 0) return [];

  const cheapest = Math.min(...bids.map((b) => b.amount as number));

  return bids
    .map((bid) => {
      // A **proportional premium over the cheapest bid**, not a position
      // within the spread.
      //
      // Scoring by position in the spread was the first version, and it is
      // wrong in a way that only shows up with real numbers: with two bids of
      // ₦1,800,000 and ₦2,000,000 the dearer one scores zero on price — as
      // though it were infinitely expensive — because it happens to be the
      // top of a two-bid range. That handed every load to the cheapest bidder
      // regardless of record, which is the exact failure this ranking exists
      // to avoid.
      //
      // Proportional keeps what the spread version was reaching for: ₦50,000
      // apart is a lot on a city run and nothing on a Kano haul, and a
      // percentage says so without letting the range set the sensitivity.
      const premium = cheapest === 0 ? 0 : (bid.amount - cheapest) / cheapest;
      const price = clamp(1 - premium / PREMIUM_TOLERANCE);

      const reliability =
        bid.tripsCompleted >= MINIMUM_TRIPS_FOR_RELIABILITY
          ? clamp(bid.tripsOnTime / bid.tripsCompleted)
          : null;

      const kmToPickup = distance(bid.at, pickup) / 1000;
      // Within 50 km is as good as at the door; beyond 300 km the truck is
      // unlikely to arrive when it says.
      const proximity = clamp(1 - Math.max(0, kmToPickup - 50) / 250);

      const score = clamp(
        0.4 * price + 0.4 * (reliability ?? 0.6) + 0.2 * proximity,
      );

      return {
        bid,
        score,
        reliability,
        kmToPickup: Math.round(kmToPickup),
        because:
          reliability === null
            ? `New to Backhaul — ${bid.tripsCompleted} completed trip` +
              `${bid.tripsCompleted === 1 ? '' : 's'}, no record yet.`
            : `${Math.round(reliability * 100)}% on time across ` +
              `${bid.tripsCompleted} trips.`,
      };
    })
    .sort((a, b) => b.score - a.score);
}

function clamp(value: number, low = 0, high = 1): number {
  if (!Number.isFinite(value)) return low;
  return Math.min(high, Math.max(low, value));
}

File: server/src/Backhaul.Domain/Market/Matching.cs
using Backhaul.Domain.Money;
using Backhaul.Domain.Pricing;
using Backhaul.Domain.Tracking;

namespace Backhaul.Domain.Market;

/// <summary>A load on the board.</summary>
public sealed record Load(
    Guid Id,
    Position Origin,
    Position Destination,
    double WeightTonnes,
    TruckClass Requires,
    Kobo? Offered,
    DateTimeOffset ReadyBy,
    DateTimeOffset ExpiresAt);

/// <summary>A truck, and where it is trying to get back to.</summary>
public sealed record Carrier(Position At, DateTimeOffset FreeFrom, TruckClass Truck, Position? Base);

public enum Blocker
{
    TooHeavy,
    WrongClass,
    Expired,
    CannotReach,
}

public sealed record LoadScore(
    Load Load,
    double Score,
    Blocker? Blocked,
    double DeadheadM,
    double ProgressHomeM,
    string Because);

/// <summary>One carrier's offer on a load.</summary>
public sealed record Bid(
    Guid Id,
    Guid CarrierId,
    Kobo Amount,
    int TripsCompleted,
    int TripsOnTime,
    Position At,
    DateTimeOffset PlacedAt);

public sealed record BidScore(Bid Bid, double Score, double? Reliability, int KmToPickup, string Because);

/// <summary>
/// Which load, and whose bid.
/// </summary>
/// <remarks>
/// Mirrors <c>packages/domain/src/matching.ts</c>. Both rankings return every
/// candidate, scored and explained, rather than a filtered shortlist — a
/// ranking that silently drops options is one the user cannot argue with, and
/// the first thing a haulier does with a recommendation is argue with it.
/// </remarks>
public static class Matching
{
    /// <summary>Deadhead beyond this is not worth scoring against.</summary>
    public static readonly double MaxDeadheadM = 400_000;

    /// <summary>Below this many completed trips, a carrier has no meaningful record.</summary>
    public static readonly int MinimumTripsForReliability = 5;

    /// <summary>
    /// How far above the cheapest bid a carrier may ask before price scores zero.
    /// </summary>
    /// <remarks>
    /// A quarter. A carrier asking 25% more than the cheapest offer had better
    /// be winning on record alone, and at that point the shipper should be
    /// looking at the two figures themselves rather than at a ranking.
    /// </remarks>
    public static readonly double PremiumTolerance = 0.25;

    /// <summary>
    /// Ranks available loads for a carrier, best first.
    /// </summary>
    /// <remarks>
    /// Blocked loads sort last regardless of score: a load that cannot be
    /// taken should never sit above one that can, however attractive it looks.
    /// </remarks>
    public static IReadOnlyList<LoadScore> RankLoads(
        Carrier carrier,
        IReadOnlyList<Load> loads,
        DateTimeOffset now) =>
        loads
            .Select(load => ScoreLoad(carrier, load, now))
            .OrderBy(s => s.Blocked is null ? 0 : 1)
            .ThenByDescending(s => s.Score)
            .ToList();

    private static LoadScore ScoreLoad(Carrier carrier, Load load, DateTimeOffset now)
    {
        var deadhead = Geo.Distance(carrier.At, load.Origin);
        var haul = Geo.Distance(load.Origin, load.Destination);

        var progressHome = carrier.Base is null
            ? 0
            : Geo.Distance(carrier.At, carrier.Base) - Geo.Distance(load.Destination, carrier.Base);

        var blocked = BlockerFor(carrier, load, deadhead, now);
        if (blocked is { } reason)
        {
            return new LoadScore(load, 0, reason, deadhead, progressHome, ExplainBlocker(reason));
        }

        // Three things decide it, and the weights say which matters:
        //
        //   value    what the trip pays against what it costs to reach
        //   homeward how much of the empty run home this load covers
        //   urgency  how soon it has to be collected
        //
        // Homeward is weighted almost as heavily as value on purpose. It is
        // the asymmetry the product is named after, and a matcher that treats
        // a return load as just another load is a load board.
        var value = ValueScore(load, haul, deadhead);
        var homeward = HomewardScore(progressHome, haul, carrier.Base is not null);
        var urgency = UrgencyScore(load, now);

        var score = Clamp(0.45 * value + 0.4 * homeward + 0.15 * urgency);

        return new LoadScore(
            load,
            score,
            null,
            deadhead,
            progressHome,
            Explain(deadhead, progressHome, carrier.Base is not null));
    }

    private static Blocker? BlockerFor(Carrier carrier, Load load, double deadhead, DateTimeOffset now)
    {
        if (load.ExpiresAt <= now) return Blocker.Expired;
        if (!Trucks.Fits(carrier.Truck, (decimal)load.WeightTonnes)) return Blocker.TooHeavy;
        if (carrier.Truck != load.Requires) return Blocker.WrongClass;
        if (deadhead > MaxDeadheadM) return Blocker.CannotReach;
        return null;
    }

    /// <summary>
    /// Paid distance against total distance.
    /// </summary>
    /// <remarks>
    /// A 500 km haul reached by 50 km of empty running is 91% productive. The
    /// same haul reached by 400 km of empty running is 56%, and no rate makes
    /// that up.
    /// </remarks>
    private static double ValueScore(Load load, double haul, double deadhead)
    {
        var total = haul + deadhead;
        if (total == 0) return 0;
        var productive = haul / total;

        var indicative = Quote.For(load.Requires, (int)haul).Mid;
        var premium = load.Offered is not { } offered || indicative.Value == 0
            ? 1
            : Clamp((double)offered.Value / indicative.Value, 0.5, 1.5);

        return Clamp(productive * premium);
    }

    /// <summary>
    /// How much of the run home this load covers, as a fraction of its length.
    /// </summary>
    /// <remarks>
    /// A load that goes exactly the right way scores 1. One that goes sideways
    /// scores 0.5 — neutral, not punished, because a paying sideways load
    /// still beats an empty truck. With no base it is a flat 0.5: a carrier
    /// who has not said where home is should not have every load marked down.
    /// </remarks>
    private static double HomewardScore(double progressHome, double haul, bool hasBase) =>
        !hasBase || haul == 0 ? 0.5 : Clamp(0.5 + progressHome / (2 * haul));

    /// <summary>Loads that must move today outrank loads that can wait a week.</summary>
    private static double UrgencyScore(Load load, DateTimeOffset now)
    {
        var hours = (load.ReadyBy - now).TotalHours;
        if (hours <= 0) return 1;
        if (hours >= 72) return 0;
        return 1 - hours / 72;
    }

    private static string Explain(double deadhead, double progressHome, bool hasBase)
    {
        var empty = $"{Round(deadhead / 1000)} km empty to the pickup";
        if (!hasBase) return $"{empty}.";
        if (progressHome > 50_000)
        {
            return $"{empty}, and it covers {Round(progressHome / 1000)} km of the run home.";
        }

        if (progressHome < -50_000)
        {
            return $"{empty}, but it takes you {Round(-progressHome / 1000)} km further from base.";
        }

        return $"{empty}; neither toward base nor away from it.";
    }

    private static string ExplainBlocker(Blocker blocked) => blocked switch
    {
        Blocker.TooHeavy => "Heavier than your truck carries.",
        Blocker.WrongClass => "The shipper asked for a different class of truck.",
        Blocker.Expired => "This load has expired.",
        Blocker.CannotReach => $"More than {MaxDeadheadM / 1000} km of empty running away.",
        _ => throw new InvalidOperationException($"unmapped blocker {blocked}"),
    };

    /// <summary>
    /// Ranks bids for a shipper.
    /// </summary>
    /// <remarks>
    /// The cheapest bid is not the best bid. A new carrier is scored as
    /// *unknown* rather than unreliable — which sits between good and bad
    /// rather than at the bottom, because a marketplace that never surfaces a
    /// new carrier never gets a second one.
    /// </remarks>
    public static IReadOnlyList<BidScore> RankBids(IReadOnlyList<Bid> bids, Position pickup)
    {
        if (bids.Count == 0) return [];

        var cheapest = bids.Min(b => b.Amount.Value);

        return bids
            .Select(bid =>
            {
                // A proportional premium over the cheapest bid, not a position
                // within the spread. Scoring by position was the first version
                // and it handed every load to the cheapest bidder: with two
                // bids of ₦1,800,000 and ₦2,000,000 the dearer scores zero on
                // price, as though it were infinitely expensive, purely
                // because it is the top of a two-bid range.
                var premium = cheapest == 0 ? 0 : (double)(bid.Amount.Value - cheapest) / cheapest;
                var price = Clamp(1 - premium / PremiumTolerance);

                double? reliability = bid.TripsCompleted >= MinimumTripsForReliability
                    ? Clamp((double)bid.TripsOnTime / bid.TripsCompleted)
                    : null;

                var kmToPickup = Geo.Distance(bid.At, pickup) / 1000.0;
                // Within 50 km is as good as at the door; beyond 300 km the
                // truck is unlikely to arrive when it says.
                var proximity = Clamp(1 - Math.Max(0, kmToPickup - 50) / 250);

                var score = Clamp(0.4 * price + 0.4 * (reliability ?? 0.6) + 0.2 * proximity);

                return new BidScore(
                    bid,
                    score,
                    reliability,
                    (int)Round(kmToPickup),
                    reliability is null
                        ? $"New to Backhaul — {bid.TripsCompleted} completed trip" +
                          $"{(bid.TripsCompleted == 1 ? "" : "s")}, no record yet."
                        : $"{Round(reliability.Value * 100)}% on time across " +
                          $"{bid.TripsCompleted} trips.");
            })
            .OrderByDescending(s => s.Score)
            .ToList();
    }

    /// <summary>
    /// JavaScript's <c>Math.round</c>: halves go up, not away from zero.
    /// </summary>
    /// <remarks>
    /// Every rounded figure here lands in a sentence the parity fixtures
    /// assert character for character, and .NET's default disagrees with
    /// JavaScript on a negative half — "1 km further from base" against "2".
    /// </remarks>
    private static double Round(double value) => Math.Floor(value + 0.5);

    private static double Clamp(double value, double low = 0, double high = 1)
    {
        if (double.IsNaN(value) || double.IsInfinity(value)) return low;
        return Math.Min(high, Math.Max(low, value));
    }
}
